db/firebase/parser.js

```diff
@@ -0,0 +1,101 @@
+// Firebase requirements
+require("firebase/firestore");
+const csv = require('csv-parser')
+const fs = require('fs')
+require('dotenv').config()
+const admin = require('firebase-admin');
+
+// initialize firebase
+function initializeFirebaseAdmin(adminSDK, databaseURL) {
+    return admin.initializeApp({
+        credential: admin.credential.cert(adminSDK),
+        databaseURL: databaseURL,
+    });
+}
+const adminSDK = JSON.parse(process.env.NWPLUSADMINSDK);
+let app = initializeFirebaseAdmin(adminSDK, "https://nwplus-bot.firebaseio.com");
+
+// save second argument as type of members to add
+let type = process.argv[2];
+if (type == undefined) {
+    throw new Error('no defined type!')
+}
+// optional third argument; if "true", all their previous types will be overwritten by this new one
+let overwrite = false;
+if (process.argv[3] === 'true') {
+    overwrite = true;
+};
+
+class Registration {
+    constructor(email) {
+        this.email = email
+        this.types = [];
+    }
+}
+
+const results = []
+const all_regs = {}
+fs.createReadStream('registrations.csv') // requires a registrations.csv file in root directory to run
+    .pipe(csv())
+    .on('data', (data) => results.push(data))
+    .on('end', () => {
+        const processed = results.map((row) => {
+            const email = row['What is your primary email that can we contact you with?'] || row['Email Address']
+
+            const r = new Registration(email)
+            return r
+        })
+
+        processed.forEach(r => {
+            all_regs[r.email] = r
+        })
+
+
+        var db = app.firestore();
+
+        var all = db.collection("members").get().then(snapshot => {
+            // get all ids and types of members already in collections and store in idMap
+            let ids = snapshot.docs.map(doc => doc.id)
+            let types = snapshot.docs.map(doc => doc.get('types'))
+            let idMap = new Map(); // Map<string, array<Type>> where Type is an object consisting of the fields type and isVerified
+            var i = 0;
+            ids.forEach(id => {
```

Comment:
I feel like there are a lot of extra maps and forEach loops here. Could you just do one forEach in snapshots.docs and from there get the idMap. Without having to use ids and types? @mwang2000 

Uses of ids.length can be replaced with idMaps.size since they are both the same length(size)

Reply:
Good point. That is doable in one forEach.